warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java

```diff
@@ -0,0 +1,234 @@
+package datawave.query.index.lookup;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+import datawave.query.jexl.JexlNodeFactory;
+import datawave.query.jexl.nodes.ExceededTermThresholdMarkerJexlNode;
+import datawave.query.jexl.nodes.ExceededValueThresholdMarkerJexlNode;
+import datawave.query.jexl.nodes.IndexHoleMarkerJexlNode;
+import datawave.query.jexl.visitors.TreeFlatteningRebuildingVisitor;
+import org.apache.commons.jexl2.parser.ASTDelayedPredicate;
+import org.apache.commons.jexl2.parser.ASTOrNode;
+import org.apache.commons.jexl2.parser.ASTReference;
+import org.apache.commons.jexl2.parser.ASTReferenceExpression;
+import org.apache.commons.jexl2.parser.JexlNode;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static datawave.query.jexl.JexlASTHelper.nodeToKey;
+
+/**
+ * This class generates the union of two {@link IndexInfo} objects. It handles all combinations of document and infinite ranges and will add in any nodes that
+ * were delayed. These are nodes that the DefaultQueryPlanner determined were not appropriate for the index lookup, thus delaying them for later evaluation.
+ * <p>
+ * Consider the following example, where the query 'A OR B' is executed against the index. Suppose that term A hits on documents 1 and 2, and term b hits on
+ * documents 2 and 3. Logically, we can represent this as,
+ * <p>
+ * <code>
+ * A = {doc1, doc2}
+ * B = {doc2, doc3}
+ * </code>
+ * <p>
+ * This is represented in the code as IndexInfo objects with an IndexMatch object for each hit in the index.
+ * <p>
+ * <code>
+ * IndexInfo(A)
+ * doc1 = {A}
+ * doc2 = {A}
+ * 
+ * IndexInfo(B)
+ * doc2 = {B}
+ * doc3 = {B}
+ * </code>
+ * <p>
+ * The union of IndexInfo(A) and IndexInfo(B) produce the following,
+ * <p>
+ * <code>
+ * IndexInfo(A OR B)
+ * doc 1 = {A}
+ * doc 2 = {A, B}
+ * doc 3 = {B}
+ * </code>
+ */
+public class IndexInfoUnion {
```

Comment:
Hmmmm, why did this need to be broken out into a separate class?  This makes tracking your changes a bit harder than it would be otherwise.  Plus, if we ever decided to extend IndexInfo, we wouldn't be able to extend these methods.

Reply:
The original implementation was hard to follow and read like a utility method (taking another IndexInfo, returning a brand new one). I broke the union operation out for the refactor and decided to leave it out, a la the Single Responsibility Principle.

Reply:
I tend to agree with @apmoriarty on this one, but I feel like the naming feels funny to me. Not sure I like IndexInfoUnion as this class does not represent an IndexInfoUnion, it merely contains the method to generate one.